Ledger does a dirty shutdown on SIGTERM since 0.3.0

Since 0.3.0, ledger state always gets corrupted when the `anoman` process is killed with `SIGTERM` (which is the default signal e.g. systemd sends to stop a service). To start the node again after `SIGTERM` it's necessary to reset the state - a panic occurs at startup otherwise.

Using `SIGINT` instead (e.g. what CTRL-C does) does do a clean shutdown, so I've been using that as a workaround with systemd (`KillSignal=SIGINT` in the service definition).

`SIGTERM` used to do a clean shutdown in 0.2.0. It would be nice to get back that behavior, since it is what is expected per POSIX.

Possibly related, this is what happens when `anoman` gets `SIGTERM` when it just recently started syncing the chain:

```
E[2022-01-04|15:36:28.252] Stopping abci.socketClient for error: read message: EOF module=abci-client connection=mempool
I[2022-01-04|15:36:28.252] Stopping socketClient service                module=abci-client connection=mempool impl=socketClient
E[2022-01-04|15:36:28.252] mempool connection terminated. Did the application crash? Please restart tendermint module=proxy err="read message: EOF"
E[2022-01-04|15:36:28.253] Stopping abci.socketClient for error: read message: EOF module=abci-client connection=consensus
I[2022-01-04|15:36:28.253] Stopping socketClient service                module=abci-client connection=consensus impl=socketClient
Terminated
E[2022-01-04|15:36:28.253] error in proxyAppConn.EndBlock               module=state err="read message: EOF"

panic: Failed to process committed block (444:6B813EC7C984A6FFC6ABED78634783D69C92A722B931510680BF2743959770EE): read message: EOF

goroutine 5401 [running]:github.com/tendermint/tendermint/blockchain/v0.(*BlockchainReactor).poolRoutine(0xc000230700, 0x0)
        github.com/tendermint/tendermint/blockchain/v0/reactor.go:401 +0x15bf
created by github.com/tendermint/tendermint/blockchain/v0.(*BlockchainReactor).OnStart
        github.com/tendermint/tendermint/blockchain/v0/reactor.go:110 +0x8c
```

It still starts fine again after that, but the sync restarts from the first block instead of where it left off. Meanwhile `SIGINT` does not produce any errors and sync continues from the block where it left off.

--- a/.changelog/unreleased/bug-fixes/768-768.md
+++ b/.changelog/unreleased/bug-fixes/768-768.md
@@ -1,0 +1,2 @@
+- Ledger: Handle Unix and Windows interrupt and termination signals to shut down
+  cleanly. ([#768](https://github.com/anoma/anoma/issues/768))

--- a/apps/src/lib/node/ledger/mod.rs
+++ b/apps/src/lib/node/ledger/mod.rs
@@ -463,10 +463,10 @@ impl Drop for Aborter {
 async fn wait_for_abort(
     mut abort_recv: tokio::sync::mpsc::UnboundedReceiver<&'static str>,
 ) {
-    let mut sigterm = tokio::signal::unix::signal(
-        tokio::signal::unix::SignalKind::terminate(),
-    )
-    .unwrap();
+    use tokio::signal::unix::{signal, SignalKind};
+    let mut sigterm = signal(SignalKind::terminate()).unwrap();
+    let mut sighup = signal(SignalKind::hangup()).unwrap();
+    let mut sigpipe = signal(SignalKind::pipe()).unwrap();
     let _ = tokio::select! {
         signal = tokio::signal::ctrl_c() => {
             match signal {
@@ -478,6 +478,18 @@ async fn wait_for_abort(
             match signal {
                 Some(()) => tracing::info!("Received termination signal, exiting..."),
                 None => tracing::error!("Termination signal cannot be caught anymore, exiting..."),
+            }
+        },
+        signal = sighup.recv() => {
+            match signal {
+                Some(()) => tracing::info!("Received hangup signal, exiting..."),
+                None => tracing::error!("Hangup signal cannot be caught anymore, exiting..."),
+            }
+        },
+        signal = sigpipe.recv() => {
+            match signal {
+                Some(()) => tracing::info!("Received pipe signal, exiting..."),
+                None => tracing::error!("Pipe signal cannot be caught anymore, exiting..."),
             }
         },
         msg = abort_recv.recv() => {

--- a/apps/src/lib/node/ledger/mod.rs
+++ b/apps/src/lib/node/ledger/mod.rs
@@ -290,7 +290,7 @@ async fn run_aux(config: config::Ledger, wasm_dir: PathBuf) {
     let tendermint_config = config.tendermint.clone();
 
     // Channel for signalling shut down from the shell or from Tendermint
-    let (abort_send, mut abort_recv) =
+    let (abort_send, abort_recv) =
         tokio::sync::mpsc::unbounded_channel::<&'static str>();
 
     // Channel for signalling shut down to Tendermint process
@@ -361,21 +361,7 @@ async fn run_aux(config: config::Ledger, wasm_dir: PathBuf) {
         .expect("Must be able to start a thread for the shell");
 
     // Wait for interrupt signal or abort message
-    tokio::select! {
-        signal = tokio::signal::ctrl_c() => {
-            match signal {
-                Ok(()) => tracing::info!("Received interrupt signal, exiting..."),
-                Err(err) => tracing::error!("Failed to listen for CTRL+C signal: {}", err),
-            }
-        },
-        msg = abort_recv.recv() => {
-            // When the msg is `None`, there are no more abort senders, so both
-            // Tendermint and the shell must have already exited
-            if let Some(who) = msg {
-                 tracing::info!("{} has exited, shutting down...", who);
-            }
-        }
-    };
+    wait_for_abort(abort_recv).await;
 
     // Abort the ABCI service task
     abci.abort();
@@ -471,4 +457,56 @@ impl Drop for Aborter {
         // Send abort message, ignore result
         let _ = self.sender.send(self.who);
     }
+}
+
+#[cfg(unix)]
+async fn wait_for_abort(
+    mut abort_recv: tokio::sync::mpsc::UnboundedReceiver<&'static str>,
+) {
+    let mut sigterm = tokio::signal::unix::signal(
+        tokio::signal::unix::SignalKind::terminate(),
+    )
+    .unwrap();
+    let _ = tokio::select! {
+        signal = tokio::signal::ctrl_c() => {
+            match signal {
+                Ok(()) => tracing::info!("Received interrupt signal, exiting..."),
+                Err(err) => tracing::error!("Failed to listen for CTRL+C signal: {}", err),
+            }
+        },
+        signal = sigterm.recv() => {
+            match signal {
+                Some(()) => tracing::info!("Received termination signal, exiting..."),
+                None => tracing::error!("Termination signal cannot be caught anymore, exiting..."),
+            }
+        },
+        msg = abort_recv.recv() => {
+            // When the msg is `None`, there are no more abort senders, so both
+            // Tendermint and the shell must have already exited
+            if let Some(who) = msg {
+                 tracing::info!("{} has exited, shutting down...", who);
+            }
+        }
+    };
+}
+
+#[cfg(not(unix))]
+async fn wait_for_abort(
+    mut abort_recv: tokio::sync::mpsc::UnboundedReceiver<&'static str>,
+) {
+    let _ = tokio::select! {
+        signal = tokio::signal::ctrl_c() => {
+            match signal {
+                Ok(()) => tracing::info!("Received interrupt signal, exiting..."),
+                Err(err) => tracing::error!("Failed to listen for CTRL+C signal: {}", err),
+            }
+        },
+        msg = abort_recv.recv() => {
+            // When the msg is `None`, there are no more abort senders, so both
+            // Tendermint and the shell must have already exited
+            if let Some(who) = msg {
+                 tracing::info!("{} has exited, shutting down...", who);
+            }
+        }
+    };
 }

--- a/apps/src/lib/node/ledger/mod.rs
+++ b/apps/src/lib/node/ledger/mod.rs
@@ -502,7 +502,35 @@ async fn wait_for_abort(
     };
 }
 
-#[cfg(not(unix))]
+#[cfg(windows)]
+async fn wait_for_abort(
+    mut abort_recv: tokio::sync::mpsc::UnboundedReceiver<&'static str>,
+) {
+    let mut sigbreak = tokio::signal::windows::ctrl_break().unwrap();
+    let _ = tokio::select! {
+        signal = tokio::signal::ctrl_c() => {
+            match signal {
+                Ok(()) => tracing::info!("Received interrupt signal, exiting..."),
+                Err(err) => tracing::error!("Failed to listen for CTRL+C signal: {}", err),
+            }
+        },
+        signal = sigbreak.recv() => {
+            match signal {
+                Some(()) => tracing::info!("Received break signal, exiting..."),
+                None => tracing::error!("Break signal cannot be caught anymore, exiting..."),
+            }
+        },
+        msg = abort_recv.recv() => {
+            // When the msg is `None`, there are no more abort senders, so both
+            // Tendermint and the shell must have already exited
+            if let Some(who) = msg {
+                 tracing::info!("{} has exited, shutting down...", who);
+            }
+        }
+    };
+}
+
+#[cfg(not(any(unix, windows)))]
 async fn wait_for_abort(
     mut abort_recv: tokio::sync::mpsc::UnboundedReceiver<&'static str>,
 ) {

